codalab/bin/ws_server.py

```diff
@@ -1,75 +1,143 @@
 # Main entry point for CodaLab cl-ws-server.
 import argparse
 import asyncio
+from collections import defaultdict
 import logging
+import os
 import re
-from typing import Any, Dict
+from typing import Any, Dict, Optional
 import websockets
+from dataclasses import dataclass
+import threading
 
-logger = logging.getLogger(__name__)
-logger.setLevel(logging.WARNING)
-logging.basicConfig(format='%(asctime)s %(message)s %(pathname)s %(lineno)d')
-
-worker_to_ws: Dict[str, Any] = {}
+from codalab.lib.codalab_manager import CodaLabManager
 
 
-async def rest_server_handler(websocket):
-    """Handles routes of the form: /main. This route is called by the rest-server
-    whenever a worker needs to be pinged (to ask it to check in). The body of the
-    message is the worker id to ping. This function sends a message to the worker
-    with that worker id through an appropriate websocket.
+@dataclass
+class WS:
```

Comment:
I might want to give this a better name. It basically just stores a lock associated with the websocket now (to indicate that that socket is already in use). Previously, it did more.